api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -276,4 +359,8 @@ message Resource {
   // Cache control properties for the resource.
   // [#not-implemented-hide:]
   CacheControl cache_control = 7;
+
+  // Dynamic parameter constraints associated with this resource. To be used by client-side caches
+  // (including xDS proxies) when matching subscribed resource locators.
+  repeated DynamicParameterConstraints dynamic_parameter_constraints = 8;
```

Comment:
> Naively I'd expect the client to say "I'll accept any patch version between [0, 15]". The server then delivers some resource at patch version 14. Instead, here we say "I'll only accept patch version 16, then some resource which says "I'll match requests for version [0, 16]" is delivered.

Yes, that's essentially correct.  I would think of the client's request not as saying "I'll only accept patch version 16" but rather "I support patch version 16", and the server's response is essentially "This resource works for versions in the range [0,16]".

> Playing devils advocate, I'm wondering which situations this breaks down. For example, let's say there are 1M resources each with some id field numbered incrementally from [0,1M). The client is willing to accept any even resource (imagine there is some range constraint that can express this). This doesn't work in this scheme, as the client must pick a specific resource, e.g. 42, but the cache then must hold 500k resources potentially rather than just 1. Have I understood correctly or is this contrived?

You're correct that the cache behavior would be very sub-optimal in that situation.  However, I can't think of a realistic use-case like that where you wouldn't want to structure it some other way.  I think that in any case where there are a large number of unique resources, you'd need to either (a) have the clients explicitly indicate which resource they want, which would wind up looking like the "Sharding Clusters" use-case in the design doc or (b) have the server decide which resource each client should use, which would wind up looking like the "Sharding Endpoints" use-case in the design doc.

Also, it's worth noting that even if there was such a use-case, it would probably be a very rare case, so it may not make sense to worry so much about optimizing it.  In general, the reason why I flipped the constraints to the resource and the parameters to the client is that I think this approach yields much less cache pollution in the common case.  In general, I think servers will know which set of clients a given resource is appropriate for based on their parameters, so they can express the same mapping with fewer cache entries.